Caused by: java.lang.IllegalArgumentException: Cannot subclass primitive, array or final types

I have a base model in kotlin similar to the following:
```
@MappedSuperclass
abstract class BaseModel : Serializable {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null
    var deleted: Boolean = false
}
```

And another `entity` inheriting from the base model similar to:
```
@Entity
data class Person(
        var firstName: String = "",
        var lastName: String = "",
        @get:Email
        var email: String = ""
) : BaseModel()
```
Stack trace:
```
 ERROR [io.qua.dev.DevModeMain] (main) Failed to start Quarkus: java.lang.ExceptionInInitializerError
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:398)
	at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:80)
	at io.quarkus.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:60)
	at io.quarkus.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:236)
	at io.quarkus.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:39)
	at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:131)
	at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:84)
	at io.quarkus.dev.DevModeMain.start(DevModeMain.java:104)
	at io.quarkus.dev.DevModeMain.main(DevModeMain.java:46)
Caused by: java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:482)
	... 10 more
Caused by: java.lang.IllegalArgumentException: Could not create type
	at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:154)
	at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:365)
Caused by: java.lang.RuntimeException: Failed to start quarkus

	at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:174)
Caused by: java.lang.IllegalArgumentException: Could not create type

	at net.bytebuddy.TypeCache$WithInlineExpunction.findOrInsert(TypeCache.java:376)
	at org.hibernate.bytecode.internal.bytebuddy.ByteBuddyState.load(ByteBuddyState.java:175)
	at org.hibernate.bytecode.internal.bytebuddy.ByteBuddyState.loadProxy(ByteBuddyState.java:99)
	at org.hibernate.proxy.pojo.bytebuddy.ByteBuddyProxyHelper.buildProxy(ByteBuddyProxyHelper.java:54)
	at io.quarkus.hibernate.orm.runtime.proxies.ProxyDefinitions.generateProxyClass(ProxyDefinitions.java:85)
	at io.quarkus.hibernate.orm.runtime.proxies.ProxyDefinitions.createFromMetadata(ProxyDefinitions.java:47)
	at io.quarkus.hibernate.orm.runtime.boot.FastBootMetadataBuilder.build(FastBootMetadataBuilder.java:338)
	at io.quarkus.hibernate.orm.runtime.PersistenceUnitsHolder.createMetadata(PersistenceUnitsHolder.java:111)
	at io.quarkus.hibernate.orm.runtime.PersistenceUnitsHolder.constructMetadataAdvance(PersistenceUnitsHolder.java:84)
	at io.quarkus.hibernate.orm.runtime.PersistenceUnitsHolder.initializeJpa(PersistenceUnitsHolder.java:47)
	at io.quarkus.hibernate.orm.runtime.HibernateOrmRecorder$4.created(HibernateOrmRecorder.java:75)
	at io.quarkus.arc.runtime.ArcRecorder.initBeanContainer(ArcRecorder.java:102)
	at io.quarkus.deployment.steps.ArcProcessor$generateResources28.deploy_0(ArcProcessor$generateResources28.zig:454)
	at io.quarkus.deployment.steps.ArcProcessor$generateResources28.deploy(ArcProcessor$generateResources28.zig:36)
	at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:444)
	... 10 more
Caused by: java.lang.IllegalArgumentException: Cannot subclass primitive, array or final types: class ...
	at net.bytebuddy.ByteBuddy.subclass(ByteBuddy.java:406)
Caused by: java.lang.IllegalArgumentException: Cannot subclass primitive, array or final types: class ...

	at net.bytebuddy.ByteBuddy.subclass(ByteBuddy.java:299)
	at org.hibernate.proxy.pojo.bytebuddy.ByteBuddyProxyHelper.lambda$buildProxy$0(ByteBuddyProxyHelper.java:57)
	at org.hibernate.bytecode.internal.bytebuddy.ByteBuddyState.lambda$load$0(ByteBuddyState.java:178)
	at net.bytebuddy.TypeCache.findOrInsert(TypeCache.java:152)
	... 27 more
```

Failing versions: 1.3.0.CR1 and CR2
Prev working versions: 1.2.0 and 1.2.1

--- a/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/proxies/ProxyDefinitions.java
+++ b/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/proxies/ProxyDefinitions.java
@@ -1,5 +1,7 @@
 package io.quarkus.hibernate.orm.runtime.proxies;
 
+import static org.objectweb.asm.Opcodes.ACC_FINAL;
+
 import java.lang.reflect.Constructor;
 import java.util.Collections;
 import java.util.HashMap;
@@ -14,6 +16,7 @@ import org.hibernate.internal.util.collections.ArrayHelper;
 import org.hibernate.mapping.PersistentClass;
 import org.hibernate.proxy.pojo.ProxyFactoryHelper;
 import org.hibernate.proxy.pojo.bytebuddy.ByteBuddyProxyHelper;
+import org.jboss.logging.Logger;
 
 /**
  * Runtime proxies are used by Hibernate ORM to handle a number of corner cases;
@@ -25,10 +28,22 @@ import org.hibernate.proxy.pojo.bytebuddy.ByteBuddyProxyHelper;
  * Implementors of a custom {@link org.hibernate.bytecode.spi.ProxyFactoryFactory} are
  * then able to lookup such class definitions at runtime to create new instances of the
  * required enhanced proxies.
+ *
+ * Failure to generate such a proxy is not critical, but it implies that Hibernate ORM
+ * will not be able to use the enhanced proxy mechanism, possibly having to generate
+ * an additional round trip to the database in some circumstances.
+ * Most notably we'll fail to generate such a proxy when the entity has a "final" modifier;
+ * we'll also need a default constructor.
+ * Default constructors are required beyond proxy generation, so a lack of such a constructor
+ * will have us abort the bootstrap process with a critical error.
+ * On the other hand, having the entities marked as "final" is handled gracefully, as we
+ * can simply fallback to not use the enhanced proxy for the specific entity, and because
+ * it's a common case when writing entities in Kotlin.
  */
 public final class ProxyDefinitions {
 
     private final Map<Class<?>, ProxyClassDetailsHolder> proxyDefinitionMap;
+    private static final Logger LOGGER = Logger.getLogger(ProxyDefinitions.class.getName());
 
     private ProxyDefinitions(Map<Class<?>, ProxyClassDetailsHolder> proxyDefinitionMap) {
         this.proxyDefinitionMap = proxyDefinitionMap;
@@ -45,6 +60,9 @@ public final class ProxyDefinitions {
                     if (needsProxyGeneration(persistentClass)) {
                         final Class mappedClass = persistentClass.getMappedClass();
                         final Class proxyClassDefinition = generateProxyClass(persistentClass, byteBuddyProxyHelper);
+                        if (proxyClassDefinition == null) {
+                            continue;
+                        }
                         final boolean overridesEquals = ReflectHelper.overridesEquals(mappedClass);
                         try {
                             proxyDefinitionMap.put(mappedClass,
@@ -81,6 +99,12 @@ public final class ProxyDefinitions {
     private static Class generateProxyClass(PersistentClass persistentClass, ByteBuddyProxyHelper byteBuddyProxyHelper) {
         final String entityName = persistentClass.getEntityName();
         final Class mappedClass = persistentClass.getMappedClass();
+        if ((mappedClass.getModifiers() & ACC_FINAL) == ACC_FINAL) {
+            LOGGER.warn("Could not generate an enhanced proxy for entity '" + entityName + "' (class='"
+                    + mappedClass.getCanonicalName()
+                    + "') as it's final. Your application might perform better if we're allowed to extend it.");
+            return null;
+        }
         final Set<Class> proxyInterfaces = ProxyFactoryHelper.extractProxyInterfaces(persistentClass, entityName);
         Class proxyDef = byteBuddyProxyHelper.buildProxy(mappedClass, toArray(proxyInterfaces));
         return proxyDef;

--- a/integration-tests/jpa-h2/src/main/java/io/quarkus/it/jpa/h2/CompanyCustomer.java
+++ b/integration-tests/jpa-h2/src/main/java/io/quarkus/it/jpa/h2/CompanyCustomer.java
@@ -2,7 +2,13 @@ package io.quarkus.it.jpa.h2;
 
 import javax.persistence.Entity;
 
+/**
+ * This entity is intentionally marked as final:
+ * it implies that ByteBuddy will not be allowed to extend it,
+ * and generate an enhanced proxy.
+ * We need to test for this to be handled gracefully.
+ */
 @Entity
-public class CompanyCustomer extends Customer {
+public final class CompanyCustomer extends Customer {
     String companyname;
 }

